File: tests/matcher.test.ts
import {
  CrossPlayPreference,
  Platform,
  canJoinLobby,
  createConsoleOnlyPlayer,
  filterConsoleOnlyPool,
  formLobby,
  formSquads,
  isConsoleOnlyLobby,
  platformsCompatible,
  validateParty,
} from "../src/index.js";
import type { Party, Player } from "../src/types.js";

describe("platformsCompatible", () => {
  it("allows Xbox and PlayStation under console_only", () => {
    expect(
      platformsCompatible(
        Platform.XBOX_SERIES,
        CrossPlayPreference.CONSOLE_ONLY,
        Platform.PS5,
      ),
    ).toBe(true);
    expect(
      platformsCompatible(
        Platform.PS4,
        CrossPlayPreference.CONSOLE_ONLY,
        Platform.XBOX_ONE,
      ),
    ).toBe(true);
  });

  it("blocks PC under console_only", () => {
    expect(
      platformsCompatible(
        Platform.XBOX_SERIES,
        CrossPlayPreference.CONSOLE_ONLY,
        Platform.PC,
      ),
    ).toBe(false);
    expect(
      platformsCompatible(
        Platform.PC,
        CrossPlayPreference.ALL,
        Platform.PS5,
      ),
    ).toBe(true);
  });

  it("enforces same_platform when selected", () => {
    expect(
      platformsCompatible(
        Platform.PS5,
        CrossPlayPreference.SAME_PLATFORM,
        Platform.XBOX_SERIES,
      ),
    ).toBe(false);
    expect(
      platformsCompatible(
        Platform.PS5,
        CrossPlayPreference.SAME_PLATFORM,
        Platform.PS4,
      ),
    ).toBe(true);
  });
});

describe("filterConsoleOnlyPool", () => {
  it("removes PC players and keeps consoles", () => {
    const players: Player[] = [
      createConsoleOnlyPlayer("x1", Platform.XBOX_SERIES),
      createConsoleOnlyPlayer("p1", Platform.PS5),
      { id: "pc1", platform: Platform.PC, crossPlayPreference: CrossPlayPreference.ALL },
    ];

    const result = filterConsoleOnlyPool(players);
    expect(result.accepted).toHaveLength(2);
    expect(result.rejected).toHaveLength(1);
    expect(result.rejected[0].playerId).toBe("pc1");
  });
});

describe("canJoinLobby", () => {
  const xbox = createConsoleOnlyPlayer("x1", Platform.XBOX_SERIES, { region: "na-east" });
  const ps5 = createConsoleOnlyPlayer("p1", Platform.PS5, { region: "na-east" });
  const pc = { id: "pc1", platform: Platform.PC, crossPlayPreference: CrossPlayPreference.ALL, region: "na-east" };

  it("allows console cross-play in the same lobby", () => {
    expect(canJoinLobby([xbox], ps5).compatible).toBe(true);
  });

  it("rejects PC in a console lobby", () => {
    expect(canJoinLobby([xbox], pc).compatible).toBe(false);
  });
});

describe("validateParty", () => {
  it("rejects parties with PC members for console queue", () => {
    const players = new Map<string, Player>([
      ["x1", createConsoleOnlyPlayer("x1", Platform.XBOX_SERIES)],
      ["pc1", { id: "pc1", platform: Platform.PC, crossPlayPreference: CrossPlayPreference.ALL }],
    ]);

    const party: Party = { id: "party1", memberIds: ["x1", "pc1"] };
    expect(validateParty(party, players).compatible).toBe(false);
  });

  it("allows mixed Xbox + PlayStation party", () => {
    const players = new Map<string, Player>([
      ["x1", createConsoleOnlyPlayer("x1", Platform.XBOX_SERIES)],
      ["p1", createConsoleOnlyPlayer("p1", Platform.PS5)],
    ]);

    const party: Party = { id: "party1", memberIds: ["x1", "p1"] };
    expect(validateParty(party, players).compatible).toBe(true);
  });
});

describe("formLobby", () => {
  it("builds a PC-free lobby with Xbox and PlayStation", () => {
    const players: Player[] = [
      createConsoleOnlyPlayer("x1", Platform.XBOX_SERIES, { region: "eu-west" }),
      createConsoleOnlyPlayer("p1", Platform.PS5, { region: "eu-west" }),
      createConsoleOnlyPlayer("p2", Platform.PS4, { region: "eu-west" }),
      { id: "pc1", platform: Platform.PC, crossPlayPreference: CrossPlayPreference.ALL, region: "eu-west" },
    ];

    const { lobby } = formLobby(players);
    expect(isConsoleOnlyLobby(lobby)).toBe(true);
    expect(lobby).toHaveLength(3);
    expect(lobby.some((p) => p.platform === Platform.PC)).toBe(false);
  });
});

describe("formSquads", () => {
  it("groups compatible console players into squads", () => {
    const players: Player[] = [
      createConsoleOnlyPlayer("x1", Platform.XBOX_SERIES, { skillRating: 1000, region: "na-east" }),
      createConsoleOnlyPlayer("p1", Platform.PS5, { skillRating: 1050, region: "na-east" }),
    ];

    const { squads } = formSquads(players, { squadSize: 2, lobbySize: 150, requireSameRegion: true, skillBand: 200 });
    expect(squads).toHaveLength(1);
    expect(squads[0]).toHaveLength(2);
  });
});

File: tests/matcher_test.cpp
#include <gtest/gtest.h>

#include <unordered_map>

#include "warzone/matcher.hpp"
#include "warzone/platform.hpp"
#include "warzone/types.hpp"

using namespace warzone;

TEST(PlatformsCompatible, AllowsConsoleCrossPlay) {
  EXPECT_TRUE(platformsCompatible(Platform::XboxSeries, CrossPlayPreference::ConsoleOnly,
                                  Platform::Ps5));
  EXPECT_TRUE(platformsCompatible(Platform::Ps4, CrossPlayPreference::ConsoleOnly,
                                  Platform::XboxOne));
}

TEST(PlatformsCompatible, BlocksPcUnderConsoleOnly) {
  EXPECT_FALSE(platformsCompatible(Platform::XboxSeries, CrossPlayPreference::ConsoleOnly,
                                   Platform::Pc));
  EXPECT_TRUE(platformsCompatible(Platform::Pc, CrossPlayPreference::All, Platform::Ps5));
}

TEST(PlatformsCompatible, EnforcesSamePlatform) {
  EXPECT_FALSE(platformsCompatible(Platform::Ps5, CrossPlayPreference::SamePlatform,
                                   Platform::XboxSeries));
  EXPECT_TRUE(platformsCompatible(Platform::Ps5, CrossPlayPreference::SamePlatform,
                                  Platform::Ps4));
}

TEST(FilterConsoleOnlyPool, RemovesPcPlayers) {
  const std::vector<Player> players = {
      createConsoleOnlyPlayer("x1", Platform::XboxSeries),
      createConsoleOnlyPlayer("p1", Platform::Ps5),
      Player{.id = "pc1", .platform = Platform::Pc, .crossPlayPreference = CrossPlayPreference::All},
  };

  const MatchResult result = filterConsoleOnlyPool(players);
  EXPECT_EQ(result.accepted.size(), 2u);
  EXPECT_EQ(result.rejected.size(), 1u);
  EXPECT_EQ(result.rejected.front().playerId, "pc1");
}

TEST(CanJoinLobby, AllowsConsoleCrossPlay) {
  const Player xbox = createConsoleOnlyPlayer("x1", Platform::XboxSeries, Player{.region = "na-east"});
  const Player ps5 = createConsoleOnlyPlayer("p1", Platform::Ps5, Player{.region = "na-east"});
  EXPECT_TRUE(canJoinLobby({xbox}, ps5).compatible);
}

TEST(CanJoinLobby, RejectsPcInConsoleLobby) {
  const Player xbox = createConsoleOnlyPlayer("x1", Platform::XboxSeries, Player{.region = "na-east"});
  const Player pc = Player{
      .id = "pc1",
      .platform = Platform::Pc,
      .crossPlayPreference = CrossPlayPreference::All,
      .region = std::string("na-east"),
  };
  EXPECT_FALSE(canJoinLobby({xbox}, pc).compatible);
}

TEST(ValidateParty, RejectsPcMembers) {
  std::unordered_map<std::string, Player> players = {
      {"x1", createConsoleOnlyPlayer("x1", Platform::XboxSeries)},
      {"pc1", Player{.id = "pc1", .platform = Platform::Pc, .crossPlayPreference = CrossPlayPreference::All}},
  };

  const Party party{.id = "party1", .memberIds = {"x1", "pc1"}};
  EXPECT_FALSE(validateParty(party, players).compatible);
}

TEST(ValidateParty, AllowsMixedConsoleParty) {
  std::unordered_map<std::string, Player> players = {
      {"x1", createConsoleOnlyPlayer("x1", Platform::XboxSeries)},
      {"p1", createConsoleOnlyPlayer("p1", Platform::Ps5)},
  };

  const Party party{.id = "party1", .memberIds = {"x1", "p1"}};
  EXPECT_TRUE(validateParty(party, players).compatible);
}

TEST(FormLobby, BuildsPcFreeLobby) {
  const std::vector<Player> players = {
      createConsoleOnlyPlayer("x1", Platform::XboxSeries, Player{.region = "eu-west"}),
      createConsoleOnlyPlayer("p1", Platform::Ps5, Player{.region = "eu-west"}),
      createConsoleOnlyPlayer("p2", Platform::Ps4, Player{.region = "eu-west"}),
      Player{.id = "pc1", .platform = Platform::Pc, .crossPlayPreference = CrossPlayPreference::All, .region = "eu-west"},
  };

  const LobbyFormationResult formed = formLobby(players);
  EXPECT_TRUE(isConsoleOnlyLobby(formed.lobby));
  EXPECT_EQ(formed.lobby.size(), 3u);
}

TEST(FormSquads, GroupsCompatiblePlayers) {
  const std::vector<Player> players = {
      createConsoleOnlyPlayer("x1", Platform::XboxSeries,
                              Player{.skillRating = 1000, .region = "na-east"}),
      createConsoleOnlyPlayer("p1", Platform::Ps5,
                              Player{.skillRating = 1050, .region = "na-east"}),
  };

  MatchmakingConfig config;
  config.squadSize = 2;
  config.lobbySize = 150;
  config.requireSameRegion = true;
  config.skillBand = 200;

  const SquadFormationResult formed = formSquads(players, config);
  EXPECT_EQ(formed.squads.size(), 1u);
  EXPECT_EQ(formed.squads.front().size(), 2u);
}
